coil-core/src/androidMain/kotlin/coil3/decode/FastImageDecoderDecoder.kt

```diff
@@ -0,0 +1,194 @@
+                        isSampled = multiplier < 1
+                        if (isSampled || !options.allowInexactSize) {
+                            val targetWidth = (multiplier * srcWidth).roundToInt()
+                            val targetHeight = (multiplier * srcHeight).roundToInt()
+                            setTargetSize(targetWidth, targetHeight)
+                        }
+                    }
+
+                    // Configure any other attributes.
+                    configureImageDecoderProperties()
+                }
+            } finally {
+                imageDecoder?.close()
+                toClose.close()
+            }
+        }
+        return DecodeResult(
+            image = image,
+            isSampled = isSampled,
+        )
+    }
+
+    private fun ImageDecoder.configureImageDecoderProperties() {
+        // https://github.com/element-hq/element-android/pull/7184
+        val ignoreHardware = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
+        allocator = if (options.bitmapConfig.isHardware && !ignoreHardware) {
+            ImageDecoder.ALLOCATOR_HARDWARE
+        } else {
+            ImageDecoder.ALLOCATOR_SOFTWARE
+        }
+        memorySizePolicy = if (options.allowRgb565) {
+            ImageDecoder.MEMORY_POLICY_LOW_RAM
+        } else {
+            ImageDecoder.MEMORY_POLICY_DEFAULT
+        }
+        if (options.colorSpace != null) {
+            setTargetColorSpace(options.colorSpace)
+        }
+        isUnpremultipliedRequired = !options.premultipliedAlpha
+        postProcessor = options.animatedTransformation?.asPostProcessor()
+    }
+
+    private suspend fun wrapDrawable(baseDrawable: Drawable): Drawable {
+        if (baseDrawable !is AnimatedImageDrawable) {
+            return baseDrawable
+        }
+
+        baseDrawable.repeatCount = options.repeatCount
+
+        // Set the start and end animation callbacks if any one is supplied through the request.
+        val onStart = options.animationStartCallback
+        val onEnd = options.animationEndCallback
+        if (onStart != null || onEnd != null) {
+            // Animation callbacks must be set on the main thread.
+            withContext(Dispatchers.Main.immediate) {
+                baseDrawable.registerAnimationCallback(animatable2CallbackOf(onStart, onEnd))
+            }
+        }
+
+        // Wrap AnimatedImageDrawable in a ScaleDrawable so it always scales to fill its bounds.
+        return ScaleDrawable(baseDrawable, options.scale)
+    }
+}
+
+@RequiresApi(28)
+class FastImageDecoderFactory @JvmOverloads constructor(
+    maxParallelism: Int = DEFAULT_MAX_PARALLELISM,
+) : Decoder.Factory {
+    private val parallelismLock = Semaphore(maxParallelism)
+
+    override fun create(
+        result: SourceFetchResult,
+        options: Options,
+        imageLoader: ImageLoader,
+    ): Decoder? {
+        val mimeType = result.mimeType
+        if (mimeType == "image/gif" && NeedRewriteGifSource) return null
```

Comment:
Similar to `BitmapFactoryDecoder`, this should never return null and always handle the `result` since it'll always be the last decoder in the component registry

Reply:
ImageDecoder cannot be used all the time
When we just have a in memory stream(Source), we still need to fallback to BitmapFactory to avoid load the whole image to memory or write to temporary file